src/nupic/algorithms/TemporalMemory.cpp

```diff
@@ -268,35 +270,29 @@ static void growSynapses(Connections &connections, Random &rng, Segment segment,
   NTA_ASSERT(std::is_sorted(candidates.begin(), candidates.end()));
 
   // Remove cells that are already synapsed on by this segment
-  for (Synapse synapse : connections.synapsesForSegment(segment)) {
-    CellIdx presynapticCell =
-        connections.dataForSynapse(synapse).presynapticCell;
-    auto ineligible =
-        std::lower_bound(candidates.begin(), candidates.end(), presynapticCell);
-    if (ineligible != candidates.end() && *ineligible == presynapticCell) {
-      candidates.erase(ineligible);
+  for (const Synapse& synapse : connections.synapsesForSegment(segment)) {
+    const CellIdx presynapticCell = connections.dataForSynapse(synapse).presynapticCell;
+    const auto already = std::lower_bound(candidates.cbegin(), candidates.cend(), presynapticCell);
+    if (already != candidates.cend() && *already == presynapticCell) {
+      candidates.erase(already);
     }
   }
 
-  const UInt32 nActual =
-      std::min(nDesiredNewSynapses, (UInt32)candidates.size());
+  const size_t nActual = std::min(static_cast<size_t>(nDesiredNewSynapses), candidates.size());
 
-  // Check if we're going to surpass the maximum number of synapses.
-  const Int32 overrun =
-      (connections.numSynapses(segment) + nActual - maxSynapsesPerSegment);
+  // Check if we're going to surpass the maximum number of synapses. //TODO delegate this to createSynapse(segment)
+  const size_t overrun = (connections.numSynapses(segment) + nActual - maxSynapsesPerSegment);
   if (overrun > 0) {
-    destroyMinPermanenceSynapses(connections, rng, segment, overrun,
-                                 prevWinnerCells);
+    destroyMinPermanenceSynapses(connections, rng, segment, overrun, prevWinnerCells);
   }
 
   // Recalculate in case we weren't able to destroy as many synapses as needed.
-  const UInt32 nActualWithMax = std::min(
-      nActual, maxSynapsesPerSegment - connections.numSynapses(segment));
+  const size_t nActualWithMax = std::min(nActual, static_cast<size_t>(maxSynapsesPerSegment) - connections.numSynapses(segment));
 
   // Pick nActual cells randomly.
-  for (UInt32 c = 0; c < nActualWithMax; c++) {
-    UInt32 i = rng.getUInt32((UInt32)candidates.size());
-    connections.createSynapse(segment, candidates[i], initialPermanence);
+  for (size_t c = 0; c < nActualWithMax; c++) {
+    const auto i = rng.getUInt32(static_cast<UInt32>(candidates.size()));
+    connections.createSynapse(segment, candidates[i], initialPermanence); //TODO createSynapse create a vector of new synapses at once
```

Comment:
I disagree with this TODO note.  I don't see any advantage to making a list of synapses and then making them all at once.  It's more work & complexity to make, populate, and then iterate over an extra list.  I think the performance of this issue is likely insignificant either way.

The next line however is almost certainly a performance issue:
`candidates.erase(candidates.begin() + i);`

Reply:
> I don't see any advantage to making a list of synapses and then making them all at once. 

I tend to prefer passing data in vectors as long as possible, it avoids nested loops and is better for spotting hot-spots. Also SIMD (auto)vectorization can do a better job then. 

> The next line however is almost certainly a performance issue:
candidates.erase(candidates.begin() + i);
 
good point! 

Anyway, out of scope of the PR, updated the comments and I'll open a separate issue.

Reply:
#373